docs/proposal/2022-09-28-event-manager-extensibility.md

```diff
@@ -0,0 +1,302 @@
+```json
+[
+  {
+    "name": "Kubernetes",
+    "type": "source",
+    "description": "Kubernetes source plugin to consume events from Kubernetes Events API.",
+    "version": "v1.0.9"
+  },
+  {
+    "name": "kubectl",
+    "type": "executor",
+    "description": "Kubectl executor plugin to handle given command in configured k8s cluster.",
+    "version": "v1.0.9"
+  }
+]
+```
+
+### How can I provide plugin configuration?
+If you use terraform before, they use a notation like `<pluginname>_...` to pass settings to specific plugin. So, in Botkube, if somebody provides `KUBERNETES_CONFIG_PATH=something`, the `kubernetes` plugin executable can be run as follows.
+```go
+client := plugin.NewClient(&plugin.ClientConfig{
+    Plugins:          m.pluginMap(metadata),
+    VersionedPlugins: nil,
+    Cmd:              exec.Command(metadata.Path, "--config", "something"), // Plugin specific params goes here
+    AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
+    HandshakeConfig: plugin.HandshakeConfig{
+        ProtocolVersion:  1,
+        MagicCookieKey:   "BOTKUBE_MAGIC_COOKIE",
+        MagicCookieValue: "BOTKUBE_BASIC_PLUGIN",
+    },
+})
+```
+
+`type` can be `source` or `executor` since we decide on plugin implementation contract based on that value.
+
+### How to install/uninstall plugins?
+Once you initialize plugin, it collects plugin metadatas from plugin index and initializes all of them by using Hashicorp's plugin system.
+Each plugin is downloaded to a local cache from Github artifacts. Most probably, in Botkube there will be a configuration section to decide
+enable/disable plugins. So, plugin executable is in local cache, and disable might be removing it from plugin list, so that Botkube will not initialize it on next 
+restart. we can enable disable plugins dynamically by using Hashicorp's Go plugin API, but I haven't experimented it. At worst case, we can update configuration and restart Botkube app.
+
+### How to Upgrade Plugins?
+In Botkube configuration, we will have plugin names with their versions, so if you want to upgrade a specific version, it would be enough to change it in config and restart application.
+
+### Where is the actual plugin process?
+Plugin process is managed by Hashicorp's Go Plugin and it creates a subprocess from actual Go program. So, Botkube Go process will create another process, and Hashicorp's Go Plugin
+helps us to communicate with this sub process. In Kubernetes environment, it will be in same pod. No need to worry about size, since only the required plugins will be donwloaded
+to local cache (user's home folder/.Botkube-plugins).
+
+### Example
+You can see a working example [here](https://github.com/huseyinbabal/Botkube-plugins-playground)
+
+### Consequences
+If we agree on this design, we can extend this epic with following description to come up with new tasks.
+
+- [ ] Maintain Botkube plugins under Kubeshop organization
```

Comment:
Is there an option to just maintain them under the existing `kubeshop/botkube` repo? and once we will have a lot of them we can think of some dedicated repo.

Reply:
Here is my 2 cents:
1. Having an option to consume the plugin from an external repo is a good thing that should be supported, so we enable external (community maintained) plugins
2. Having core plugins such as kubectl, k8s and later helm, argo etc. can still stay close to the code.

Why I have issue with separate repository:
1. We will still have logic in both places in `botkube` and `botkube-plugins` repo that at the beginning will need to be synced. For example, current permission management, interactive dialogs, and similar.
2. I see a huge problem and slow down when it comes to development and testing. If we have two repos, then I need to play with both of them to do the release of plugin (put it somewhere too?) and to run botkube to consume it.
3. How we will eventually test that on CI? When a plugin is changed, will it trigger it with latest Botkube? When Botkube is changed, will it see the latest release of all plugins? What if feature must be done in both repos, how it will be tested?

**What I propose:**
1. Have an option to consume external plugins as described, but for now do it only for 3rd party plugins.
2. Core plugins still leave in dedicated folders under main `botkube` repository.

**Workflow:**
1. The plugins are built only on official Botkube release and added to the release assets.
2. On official release, the Botkube docker image doesn't have binaries included. Instead, it fetched them from a given release as described in the proposal.
3. On CI the plugins are built on change and incorporated in the Botkube image — no need to sync state between 2 repositories.
  - alternatively we can even build plugins into a single Docker and mount it as init container and copy the binaries into mounted volume (emptyDir). IMO no need to do that.
4. On local testing the plugins are built on change and incorporated in the Botkube image - no need to run it in 2 repos, no need to push it somewhere or mock download for local testing.

**Consequences:**
1. The core plugins are in the `botkube` repository (at least for the beginning)
3. Botkube can load plugins from file system without downloading bins from internet.

Reply:
AS we discussed, I am ok with that and updated doc accordingly.